pkg/member/member_add.go

```diff
@@ -0,0 +1,264 @@
+	clientFactory := etcdutil.NewFactory(etcdConn)
+	return &NewMember{
+		clientFactory: clientFactory,
+	}
+}
+
+// AddMemberAsLearner add a member as a learner to the etcd cluster
+func (m *NewMember) AddMemberAsLearner(logger *logrus.Logger) error {
+	//Add member as learner to cluster
+	logger.Info("Starting Add Member")
+	memberURL, err := getMemberURL(logger)
+	if memberURL == "" || err != nil {
+		logger.Info("Error fetching etcd member URL")
+		return fmt.Errorf("Could not fetch member URL : %v", err)
+	}
+
+	cli, err := m.clientFactory.NewCluster()
+	defer cli.Close()
+	if err != nil {
+		logger.Info("Error creating etcd client to add member")
+		return fmt.Errorf("failed to build etcd cluster client : %v", err)
+	}
+
+	memAddCtx, cancel := getContext(context.TODO(), etcdTimeout)
+	defer cancel()
+	_, err = cli.MemberAddAsLearner(memAddCtx, []string{memberURL})
+
+	if err == nil {
+		logger.Info("Added member to cluster as a learner")
+		return nil
+	}
+	if errors.Is(err, rpctypes.Error(rpctypes.ErrGRPCPeerURLExist)) {
+		logger.Info("Member already part of etcd cluster")
+		return nil
+	}
+	if err != nil {
+		logger.Warn("Error adding member as a learner: ", err)
+		return fmt.Errorf("Error adding member as a learner: %v", err)
+	}
+
+	logger.Info("Could not as member as learner due to: ", err)
+	return fmt.Errorf("Could not as member as learner due to: %v", err)
+}
+
+// IsMemberInCluster checks is the current members peer URL is already part of the etcd cluster
+func (m *NewMember) IsMemberInCluster(logger *logrus.Logger) (bool, error) {
+	// Check if an etcd is already available
+	backoff := retry.DefaultBackoff
+	backoff.Steps = 2
+	err := retry.OnError(backoff, func(err error) bool {
+		return err != nil
+	}, func() error {
+		etcdProbeCtx, cancel := getContext(context.TODO(), etcdTimeout)
+		defer cancel()
+		return miscellaneous.ProbeEtcd(etcdProbeCtx, m.clientFactory, logger)
+	})
+	if err != nil {
+		return false, err
+	}
+
+	cli, err := m.clientFactory.NewCluster()
+	if err != nil {
+		logger.Errorf("failed to build etcd cluster client")
+		return false, err
+	}
+	defer cli.Close()
+	logger.Info("Etcd client created")
+
+	// List members in cluster
+	var etcdMemberList *clientv3.MemberListResponse
+	err = retry.OnError(retry.DefaultBackoff, func(err error) bool {
+		return err != nil
+	}, func() error {
+		memListCtx, cancel := getContext(context.TODO(), etcdTimeout)
+		defer cancel()
+		etcdMemberList, err = cli.MemberList(memListCtx)
+		return err
+	})
+	if err != nil {
+		logger.Warn("Could not list any etcd members", err)
+		return false, err
+	}
+
+	podName, err := miscellaneous.GetEnvVarOrError("POD_NAME")
+	if err != nil {
+		logger.Error("Error reading POD_NAME env var", err)
+		return false, err
+	}
+	for _, y := range etcdMemberList.Members {
+		if y.Name == podName {
+			return true, err
+		}
+	}
+
+	return false, err
+}
+
+func getMemberURL(logger *logrus.Logger) (string, error) {
```

Comment:
As discussed before this is not the correct place to read the configuration file. It should only be read just once and not again and again. This method is currently called whenever `AddMemberAsListener` and `UpdateMemberPeerAddress` methods are invoked.